### ISSUE TITLE
Automatic primary key generation script

### ISSUE THREAD

**Issue by matheusmoreira** [drduh/YubiKey-Guide#105]
[I wrote a bash script][script] that automatically generates a 4096 bit RSA certification key in batch mode. The new key does not expire and has sensible algorithm and keyserver preferences. Automation can prevent mistakes and make things easy and reproducible.

Do you think something like this is a good fit for the guide? I think this is more user friendly compared to navigating menus. The current version of my script obtains user credentials from `git` but I could make it so that it prompts the user instead.

It is also possible to generate primary and subordinate keys with a few commands:

    algo=rsa4096
    uid='name <email>'
    expire=0

    gpg --batch --quick-generate-key "$uid" $algo cert $expire

    # FPR should be set to the fingerprint of the key generated above.
    # Maybe gpg --list-keys --with-colons can help automate this.

    for cap in sign encrypt auth; do
      gpg --batch --quick-add-key $FPR $algo $cap $expire
    done

[script]: https://github.com/matheusmoreira/.files/blob/master/~/.local/bin/generate-primary-key

**Comment by juanii** [comment]
I worked on something similar, taking one step further to avoid user interaction. The downside is I don't know if all the command line options are available in older OpenGPG versions. This is what I got to:

I think some gpg parameters are only available in newer versions so not everything might be suitable for this guide.

`export PASSPHRASE_FILE=$(mktemp)`

Generate and store the random passphrase:
`gpg --gen-random --armor 0 24 > "${PASSPHRASE_FILE}"`
If you're feeling insecure, you can set the umask to 0077 before this step. The file can be printed for backup and then securely deleted at the end of the process.

Create the master key:
`gpg --pinentry-mode loopback --batch --passphrase-file "${PASSPHRASE_FILE}" --quick-generate-key "_name_ <_email_>" rsa4096 cert never`

Get key ID and fingerprint for use in the following commands:
`export KEYID=$(gpg --list-keys --with-colons | grep "^pub:" | cut -d: -f5)`
`export FPR=$(gpg --list-keys --with-colons | grep "^fpr:" | cut -d: -f10)`
This contains unix-specific commands. On Windows, I think `grep` could be substituted with `find` or `findstr` and `cut` with `for`.

Create the signing, encryption and authentication subkeys:
`gpg --pinentry-mode loopback --batch --passphrase-file "${PASSPHRASE_FILE}" --quick-add-key "${FPR}" rsa4096 sign 1y`
`gpg --pinentry-mode loopback --batch --passphrase-file "${PASSPHRASE_FILE}" --quick-add-key "${FPR}" rsa4096 encr 1y`
`gpg --pinentry-mode loopback --batch --passphrase-file "${PASSPHRASE_FILE}" --quick-add-key "${FPR}" rsa4096 auth 1y`

Export the master key, subkeys and public key:
`gpg --pinentry-mode loopback --batch --passphrase-file "${PASSPHRASE_FILE}" --armor --output "${GNUPGHOME}/mastersub.key" --export-secret-keys "${KEYID}"`
`gpg --pinentry-mode loopback --batch --passphrase-file "${PASSPHRASE_FILE}" --armor --output "${GNUPGHOME}/sub.key" --export-secret-subkeys "${KEYID}"`
`gpg --pinentry-mode loopback --batch --passphrase-file "${PASSPHRASE_FILE}" --armor --output "${GNUPGHOME}/pubkey.txt" --export "${KEYID}"`
Some sites suggest adding `--export-options export-backup` which adds useful information for the restore process.

I didn't find a way to send the keys to the card without user interaction. The problem is there's no way to provide the card PIN in the command line. There's an unofficial OpenGPG patch but it's not part of the official build.

Thank you very much Dr. Duh for this guide, and I hope this contributes somehow.

Regards

**Comment by ageis** [comment]
Hey ya'll, check out this relevant ticket I opened against GnuPG: https://dev.gnupg.org/T4514

**Comment by matheusmoreira** [comment]
@ageis Thanks for opening that feature request! I was planning to work on refactoring the subkey generation code too. I decided to wait because [I already have several patches pending review][patches]. In order to test my changes, I wrote a test script that generates keys automatically:

```bash
gpg --batch --quick-gen-key test
PRIMARY=$(gpg -K --with-colons | awk -F: '/fpr/ { print $10 }' | head -1)
gpg --batch --quick-add-key $PRIMARY
```

Daniel Kahn Gillmor mentioned a workaround using `--status-fd` as well.

[patches]: https://dev.gnupg.org/T4457#125042

**Comment by drduh** [comment]
It would very helpful to automate key rotation, though I'm not so sure there's win in a script to do it only once. Please do send PR to the Rotate keys section when ready.

**Comment by ageis** [comment]
didn't catch @drduh's comment on this until now (oops, had just e-mailed them), but wanted to link my script within this issue for anyone interested, or the previous commenters @matheusmoreira / @juanii. no further activity on my GPG ticket yet unfortunately. 

[gen_smartcard_key.sh](https://gist.github.com/ageis/e609053307b5cdaed48179e111a10d67)

i'm also developing a way to generate these smartcard keys within a LUKS-encrypted alpine docker container that can run a GPG agent accessible to a remote host via OpenSSH socket forwarding, heh. will publish that when ready

**Comment by Underknowledge** [comment]
of cause I only find this repo to late. 
Just finished with a automated key setup. 
https://raw.githubusercontent.com/Underknowledge/installation-scripts/master/GPG/generate-gpg-keys.sh
`--quick-add-key` would have bean neater, but well. 

I currently  struggle to program the pins with bash. 
would have guessed something around 

```bash
(   gpg --command-fd=/dev/stdin --status-fd=/dev/stdout \
    --pinentry-mode=loopback --passphrase="$key_passphrase" \
    --card-edit
) <<EOF-cardedit
## enable admin
admin
## Enter menu to change or unblock the PIN
passwd
## 4 - set the Reset Code
4
### Fail Error setting the Reset Code: Bad PIN
${OLD_ADMIN_PIN}
${RESET_CODE}
```
but always facing `Error setting the Reset Code: Bad PIN` 
I would say it has definitely something to do with pinentry. Pointers welcome
